result/result.py

```diff
@@ -1,75 +1,137 @@
 # -*- coding: utf-8 -*-
 from __future__ import print_function, division, absolute_import, unicode_literals
+from typing import Generic, TypeVar, Union, Any, Optional, cast, overload
 
 
-class Result(object):
+E = TypeVar("E")
+T = TypeVar("T")
+A = TypeVar("A")
+
+
+class Result(Generic[E, T]):
     """
     A simple `Result` type inspired by Rust.
 
     Not all methods (https://doc.rust-lang.org/std/result/enum.Result.html)
     have been implemented, only the ones that make sense in the Python context.
-    You still don't get any type checking done.
-
     """
-    def __init__(self, force=False):
+    def __init__(self, is_ok, value, force=False):
+        # type: (bool, Union[E, T], bool) -> None
+        """Do not call this constructor, use the Ok or Err class methods instead.
+
+        There are no type guarantees on the value if this is called directly.
+
+        Args:
+            is_ok: If this represents an ok result
+            value: The value inside the result
+            force: Force creation of the object. This is false by default to prevent accidentally
+                creating instance of a Result in an unsafe way.
+        """
         if force is not True:
             raise RuntimeError("Don't instantiate a Result directly. "
                                "Use the Ok(value) and Err(error) class methods instead.")
+        else:
+            self._is_ok = is_ok
+            self._value = value
 
     def __eq__(self, other):
-        return self._type == other._type and self._val == other._val
+        # type: (Any) -> bool
+        return (self.__class__ == other.__class__ and
+                self.is_ok() == cast(Result, other).is_ok() and
+                self._value == other._value)
+
+    def __ne__(self, other):
+        # type: (Any) -> bool
+        return not (self == other)
+
+    def __hash__(self):
+        # type: () -> int
+        return hash((self.is_ok(), self._value))
+
+    def __repr__(self):
+        # type: () -> str
+        if self.is_ok():
+            return str("Ok({})").format(repr(self._value))
```

Comment:
I think this is the way to go. This should be 2-3 compatible, and I think it is in the spirit of what `repr` is supposed to do. `repr(u"£10")` is `"u'\\xa310'"` in python 2 and `"'£10'"` in python 3. Using this method `repr(Ok(u"£10"))` is `"Ok(u'\\xa310')"` in python 2 and `"Ok('£10')"` in python 3.